### PR TITLE
py/modsys: Add sys.executable.

### PR DESCRIPTION
**Pull request by jimmo**
Only supported on Unix, matches CPython. Gives the absolute path to the micropython binary.

Intended to support https://github.com/micropython/micropython-lib/pull/544

_This work was funded through GitHub Sponsors._

### CODE REVIEW

File: py/mpconfig.h [hunk 1]
```diff
@@ -1372,6 +1372,10 @@ typedef double mp_float_t;
 #define MICROPY_PY_SYS_EXC_INFO (0)
 #endif
 
+#ifndef MICROPY_PY_SYS_EXECUTABLE
```
Comment (dlech) [inline py/mpconfig.h:1375]: A comment here similar to other options would be helpful.

Comment (jimmo) [reply py/mpconfig.h:1378]: Done!

File: ports/unix/main.c [hunk 2]
```diff
@@ -435,6 +436,17 @@ STATIC void set_sys_argv(char *argv[], int argc, int start_arg) {
     }
 }
 
+#if MICROPY_PY_SYS_EXECUTABLE
+STATIC char executable_path[MICROPY_ALLOC_PATH_MAX];
+STATIC void sys_set_excecutable(char *argv0) {
+    extern mp_obj_str_t mp_sys_executable_obj;
+    if (realpath(argv0, executable_path)) {
+        mp_sys_executable_obj.data = (const byte *)executable_path;
+        mp_sys_executable_obj.len = strlen(executable_path);
```
Comment (dpgeorge) [inline ports/unix/main.c:445]: maybe also populate the hash?

File: py/modsys.c [hunk 3]
```diff
@@ -115,6 +115,11 @@ STATIC const mp_rom_obj_tuple_t mp_sys_implementation_obj = {
 STATIC const MP_DEFINE_STR_OBJ(mp_sys_platform_obj, MICROPY_PY_SYS_PLATFORM);
 #endif
 
+#ifdef MICROPY_PY_SYS_EXECUTABLE
+// executable - the path to the micropython binary
+MP_DEFINE_STR_OBJ(mp_sys_executable_obj, "");
```
Comment (dpgeorge) [inline py/modsys.c:120]: add "// This object is non-const and will be populated at start up."

**Comment by jimmo** [comment]
Done.

File: py/mpconfig.h [hunk 4]
```diff
@@ -1372,6 +1372,13 @@ typedef double mp_float_t;
 #define MICROPY_PY_SYS_EXC_INFO (0)
 #endif
 
+// Whether to provide "sys.executable", which is the absolute path to the
+// micropython binary
```
Comment (stinos) [inline py/mpconfig.h:1376]: Missing dot after binary?

**Comment by stinos** [comment]
I'd enable this by default for windows ports as well. Well actually it would be best if the general OS unix/windows ports are as close together as possible but I haven't made up my mind yet on how to achieve that. Now with the feature_level stuff it should be easier.

File: py/mpconfig.h (hunk 4 above)
Comment (jimmo) [reply py/mpconfig.h:1376]: None of the comments in this file end in full stop.

(I had to restrain myself!!)

Comment (stinos) [reply py/mpconfig.h:1376]: Oh, ok then :)

**Comment by dpgeorge** [comment]
I split the last commit into 2 (add sys.executable, then enable it on unix), and merged in c44b3927b8f464e1e01c07b2431e7db251e6d8c0 through 8e912a501a926f59f330f8a71c2ac1bd4af88a9e

A follow-up PR should enable this on the windows port.

**Pull request closed without merging**